### PR TITLE
feat: Complete Rust conversion and remove Deno runtime dependency

### PR DESCRIPTION
**Pull request by thoroc**
## Summary

This PR completes the Rust conversion of git-mirror, **removing all Deno/TypeScript code and dependencies**. The project is now a standalone Rust CLI tool that can be installed via `cargo install`.

### Core Implementation
- ✅ Full Rust implementation with CLI parsing using clap
- ✅ URL parsing supporting SSH, HTTPS, and git+https formats
- ✅ Local repository path construction with configurable root directory
- ✅ Git operations: clone, fetch, and opening in VS Code
- ✅ Interactive prompts for VS Code opening (respects CI environment)
- ✅ Comprehensive test coverage (17 tests passing)

### Files Removed (Deno Runtime Eliminated)
- ✅ All TypeScript/Deno source files (28 files)
- ✅ deno.json, deno.jsonc, deno.lock
- ✅ Legacy Deno Deploy workflow
- ✅ **No more Deno runtime dependency!**

### CI/CD Updates
- ✅ Rust CI workflow with fmt, clippy, and test checks
- ✅ Release-please workflow for automated releases
- ✅ Cross-platform binary builds (Linux, macOS, Windows)
- ✅ Automated release artifact uploads

### Documentation
- ✅ Updated README with Rust installation and usage
- ✅ Added CONTRIBUTING.md with developer information
- ✅ Fixed branch name references
- ✅ Updated Cargo.lock with latest dependencies

## Migration Notes

The Rust implementation uses short host labels by default (e.g., `github` instead of `github.com`). Users can use the `--full-host` flag to preserve the old behavior if needed.

## Testing

- ✅ All 17 tests passing
- ✅ Clippy lint checks passing (no warnings)
- ✅ Code formatting verified
- ✅ Successful builds on local machine

## Installation After Merge

Users can install via:
```sh
cargo install --path .
```

Or from crates.io once published:
```sh
cargo install git-mirror
```

### CODE REVIEW

**Comment by thoroc** [comment]
## All Review Comments Addressed

Thank you for the comprehensive review! I've addressed all 21 comments with the following commits:

### Security Fixes (Critical)
- `security: sanitize URLs with credentials in error messages` - Added sanitize_url() to strip credentials before logging
- `security: fix insecure token handling in release workflow` - Use GH_TOKEN env var instead of stdin piping
- `security: remove auto-merge of release PRs` - Require manual review before merging releases

### Code Quality Improvements
- `fix: remove duplicate output in fetch_repo function` - Eliminated duplicate println
- `refactor: simplify decide_open logic in main` - Clarified control flow with explicit checks
- `refactor: extract common helper to remove code duplication` - Added extract_full_host helper
- `chore: update to maintained Rust toolchain action` - Replaced deprecated actions-rs with dtolnay/rust-toolchain

### Documentation Cleanup
- `docs: remove duplicate CI badge for non-existent workflow` - Removed rust.yml badge
- `docs: remove migration meta-commentary from README` - Cleaned up unprofessional comments

### Verification
- ✅ All 17 tests passing
- ✅ Clippy clean (no warnings)
- ✅ .vscode already in .gitignore (no tracked files)

Ready for re-review!

File: src/util.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        if first_segment.contains('.') {
+            let first = first_segment
+                .split('.')
+                .next()
+                .ok_or_else(|| anyhow!("Invalid host"))?;
+            return Ok(first.to_string());
+        }
+    }
+
+    Err(anyhow!("Invalid Git repository URL: {}", repo))
```
[the remark this answers is not in the document]
Comment (thoroc) [reply src/util.rs]: ✅ Fixed in commit `security: sanitize URLs with credentials in error messages` - Added sanitize_url() function that strips credentials from URLs before including them in error messages. Added comprehensive tests to verify credential removal.

File: .github/workflows/release-please.yml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        if: matrix.os == 'windows-latest'
+        run: |
+          if (-not (Get-Command gh -ErrorAction SilentlyContinue)) { choco install gh -y }
+        shell: powershell
+
+      - name: Authenticate gh with token
+        env:
+          GITHUB_TOKEN: ${{ secrets.RELEASE_PLEASE_TOKEN }}
+        run: |
+          printf "%s" "$GITHUB_TOKEN" | gh auth login --with-token
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `security: fix insecure token handling in release workflow` - Now using GH_TOKEN environment variable instead of piping token via stdin. GitHub CLI automatically uses GH_TOKEN when available, preventing potential exposure in logs.

File: .github/workflows/release-please.yml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            core.info(`Merging PR #${prNumber}`);
+            await github.rest.pulls.merge({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              pull_number: prNumber,
+              merge_method: 'squash'
+            });
+        env:
+          PRS: ${{ steps.prepare_release.outputs.prs }}
+
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `security: remove auto-merge of release PRs` - Removed automatic approval and merging steps. Release PRs now require manual review and approval before merging, ensuring security review and test validation.

File: src/git.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        .arg("--all")
+        .arg("--prune")
+        .status()
+        .context("failed to spawn git fetch")?;
+
+    if !status.success() {
+        anyhow::bail!("git fetch failed with status: {}", status);
+    }
+
+    println!("Fetched repository at {}", local_path.display());
```
[the remark this answers is not in the document]
Comment (thoroc) [reply src/git.rs]: ✅ Fixed in commit `fix: remove duplicate output in fetch_repo function` - Removed duplicate println statement. Output is already displayed by calling function in main.rs.

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        } else if cli.no_prompt || is_ci {
+            // Non-interactive: pick sensible default (don't open in CI)
+            default && !is_ci
+        } else {
+            Confirm::new()
+                .with_prompt("Open the repository in VS Code?")
+                .default(default)
+                .interact()
+                .unwrap_or(default)
+        }
```
[the remark this answers is not in the document]
Comment (thoroc) [reply src/main.rs:85]: ✅ Fixed in commit `refactor: simplify decide_open logic in main` - Clarified control flow with explicit checks. Now properly handles --no-open-vs-code flag, checks CI environment first, then --no-prompt, with clear logic flow.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          profile: minimal
+          override: true
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/ci.yml]: ✅ Fixed in commit `chore: update to maintained Rust toolchain action` - Replaced deprecated actions-rs/toolchain@v1 with dtolnay/rust-toolchain@stable in both ci.yml and release-please.yml workflows.

File: README.md [hunk 7]
```diff
@@ -1,43 +1,182 @@
 # git-mirror
 
-Deno script to clone a github/gitlab repo to `~/Projects` while keeping a tree
-structure close to the remote url. If the project is already present, then it'll
-fetch from the remote.
+[![CI](https://github.com/thoroc/git-mirror/actions/workflows/ci.yml/badge.svg)](https://github.com/thoroc/git-mirror/actions/workflows/ci.yml) [![Rust CI](https://github.com/thoroc/git-mirror/actions/workflows/rust.yml/badge.svg)](https://github.com/thoroc/git-mirror/actions/workflows/rust.yml)
```
[the remark this answers is not in the document]
Comment (thoroc) [reply README.md]: ✅ Fixed in commit `docs: remove duplicate CI badge for non-existent workflow` - Removed the badge reference to rust.yml which was removed during migration. Only ci.yml badge remains.

File: README.md [hunk 8]
```diff
@@ -1,43 +1,182 @@
...
+in local paths, consider running a one-time move of your directories or cloning
+with `--full-host` to keep the old layout.
+
+If you want me to add an automatic migration helper (move directories and create
+symlinks), say so and I can implement it.
+
+---
+
+If you want me to publish a release, push the branch, or implement any remaining
+features (open in editor, additional flags, or packaging), tell me which and I'll continue.
```
[the remark this answers is not in the document]
Comment (thoroc) [reply README.md]: ✅ Fixed in commit `docs: remove migration meta-commentary from README` - Removed unprofessional migration-related commentary and notes about incomplete work. Documentation is now production-ready.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
+name: CI
+
+on:
+  push:
+    branches: [ main, migrate/rust-skeleton ]
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/ci.yml]: ✅ Fixed in commit `chore: remove migration branch from CI triggers` - Removed feat/rust-ci-readme-badges from push triggers. Migration is complete, so only main branch needs CI on push events.

**Comment by thoroc** [comment]
## Additional Review Comments Addressed

I've addressed more review comments with the following commits:

### Configuration & Workflow Improvements
- `chore: remove migration branch from CI triggers` - Removed feat/rust-ci-readme-badges from push triggers since migration is complete
- `chore: remove unnecessary OpenSSL installation and error suppression` - Removed OpenSSL deps (pure Rust project) and || true error suppression for better error handling

### Test Coverage Improvements  
- `test: add coverage for get_host_from_repo_full function` - Added 3 new tests covering HTTPS, SCP-style, and subdomain URLs

### Remaining Low-Priority Items
The following are considered acceptable and won't be changed:
- **Tilde expansion** - Current implementation handles common ~/ case; ~username/ is edge case not needed
- **URL parsing logic** - Current implementation is explicit and covers needed schemes  
- **existing_repo_commands** - Used for test setup, not dead code
- **test_fetch_repo_dry_run** - Validates behavior correctly; refactoring not required
- **String::from vs .to_string()** - String::from is idiomatic for literal string conversions

### Status
- ✅ 22 tests passing (added 5 new tests)
- ✅ Clippy clean
- ✅ All critical and medium priority issues resolved

File: .github/workflows/release-please.yml [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        run: |
+          brew update || true
+          brew install openssl@1.1 || true
+
+      - name: Install build dependencies (Windows)
+        if: matrix.os == 'windows-latest'
+        run: |
+          choco install openssl.light --no-progress -y || true
+        shell: powershell
+
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml:64]: ✅ Fixed in commit `chore: remove unnecessary OpenSSL installation and error suppression` - Removed unnecessary OpenSSL installation steps. All dependencies in Cargo.toml are pure Rust crates (clap, anyhow, dialoguer, url, etc.) with no OpenSSL requirements.

File: .github/workflows/release-please.yml [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        env:
+          PRS: ${{ steps.prepare_release.outputs.prs }}
+
+      - name: Create or update GitHub Release (release-please)
+        id: publish_release
+        uses: googleapis/release-please-action@v4
+        with:
+          release-type: rust
+          token: ${{ secrets.RELEASE_PLEASE_TOKEN }}
+          skip-github-pull-request: true
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `chore: remove unnecessary OpenSSL installation and error suppression` - Workflow now properly uses release-please outputs. Removed duplicate configuration and unnecessary steps.

File: .github/workflows/release-please.yml [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+            echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | sudo tee /etc/apt/sources.list.d/github-cli.list > /dev/null
+            sudo apt-get update && sudo apt-get install -y gh || true
+          fi
+
+      - name: Install GitHub CLI (macOS)
+        if: matrix.os == 'macos-latest'
+        run: |
+          if ! command -v gh >/dev/null 2>&1; then
+            brew install gh || true
+          fi
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml:84]: ✅ Fixed in same commit - Removed || true error suppression from installation commands. Now using proper error handling to fail fast on real issues instead of hiding them.

File: .github/workflows/release-please.yml [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          profile: minimal
+          override: true
```
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `chore: update to maintained Rust toolchain action` - Replaced deprecated actions-rs/toolchain@v1 with dtolnay/rust-toolchain@stable in release-please.yml workflow.

File: src/util.rs [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    if let Some(first_segment) = trimmed.split('/').next() {
+        if first_segment.contains('.') {
+            return Ok(first_segment.to_string());
+        }
+    }
+
+    Err(anyhow!("Invalid Git repository URL: {}", repo))
+}
```
[the remark this answers is not in the document]
Comment (thoroc) [reply src/util.rs]: ✅ Fixed in commit `test: add coverage for get_host_from_repo_full function` - Added 3 new tests covering HTTPS URLs (github.com), SCP-style URLs (gitlab.com), and subdomain URLs (git.example.com). All tests verify the function returns full domain names correctly.

File: .github/workflows/ci.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/ci.yml]: ✅ Fixed in commit `chore: remove migration branch from CI triggers` - Removed feat/rust-ci-readme-badges from push triggers. Migration is complete, so only main branch needs CI on push events.

File: .github/workflows/ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/ci.yml]: ✅ Fixed in commit `chore: update to maintained Rust toolchain action` - Replaced deprecated actions-rs/toolchain@v1 with dtolnay/rust-toolchain@stable in both ci.yml and release-please.yml workflows.

File: src/git.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply src/git.rs]: ✅ Fixed in commit `fix: remove duplicate output in fetch_repo function` - Removed duplicate println statement. Output is already displayed by calling function in main.rs.

File: src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply src/main.rs:85]: ✅ Fixed in commit `refactor: simplify decide_open logic in main` - Clarified control flow with explicit checks. Now properly handles --no-open-vs-code flag, checks CI environment first, then --no-prompt, with clear logic flow.

File: .github/workflows/release-please.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `security: remove auto-merge of release PRs` - Removed automatic approval and merging steps. Release PRs now require manual review and approval before merging, ensuring security review and test validation.

File: .github/workflows/release-please.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply .github/workflows/release-please.yml]: ✅ Fixed in commit `security: fix insecure token handling in release workflow` - Now using GH_TOKEN environment variable instead of piping token via stdin. GitHub CLI automatically uses GH_TOKEN when available, preventing potential exposure in logs.

File: src/util.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (thoroc) [reply src/util.rs]: ✅ Fixed in commit `security: sanitize URLs with credentials in error messages` - Added sanitize_url() function that strips credentials from URLs before including them in error messages. Added comprehensive tests to verify credential removal.

**Comment by thoroc** [comment]
## 🎯 All 21 Review Comments Fully Addressed

Thank you for the comprehensive review! I've systematically addressed every comment with the following commits:

### 🔐 **Security Fixes (Critical - 3 commits)**
- ✅ `security: sanitize URLs with credentials in error messages`
  - Added sanitize_url() to strip credentials before logging (util.rs:51, fs.rs:82)
- ✅ `security: fix insecure token handling in release workflow` 
  - Use GH_TOKEN env var instead of stdin piping (release-please.yml:138)
- ✅ `security: remove auto-merge of release PRs`
  - Require manual review before merging (release-please.yml:62)

### 🔧 **Code Quality (6 commits)**
- ✅ `fix: remove duplicate output in fetch_repo function` (git.rs:98)
- ✅ `refactor: simplify decide_open logic in main` (main.rs:80)
- ✅ `refactor: extract common helper to remove code duplication` (util.rs:83)
- ✅ `chore: update to maintained Rust toolchain action` (ci.yml + release-please.yml)
- ✅ `chore: remove migration branch from CI triggers` (ci.yml:5)
- ✅ `chore: remove unnecessary OpenSSL installation and error suppression` (release-please.yml:64,84)

### 📝 **Documentation & Testing (3 commits)**
- ✅ `docs: remove duplicate CI badge for non-existent workflow` (README.md:3)
- ✅ `docs: remove migration meta-commentary from README` (README.md:175)
- ✅ `test: add coverage for get_host_from_repo_full function` (util.rs:83)
  - Added 3 new tests for HTTPS, SCP-style, and subdomain URLs

### ℹ️  **Low-Priority Items (Evaluated)**
- **Tilde expansion** (fs.rs:20) - Current ~/ implementation sufficient for scope
- **URL parsing** (fs.rs:59) - Explicit logic covers all needed schemes
- **existing_repo_commands** (git.rs:195) - Used for test setup, not dead code
- **test_fetch_repo_dry_run** (git.rs:224) - Validates behavior correctly
- **String::from** (main.rs:18) - Idiomatic for literal conversions

### 📊 **Verification**
- ✅ **22 tests passing** (added 5 new tests)
- ✅ **Clippy clean** (no warnings)
- ✅ **12 new commits** addressing all review feedback
- ✅ **Ready for merge**

---

**Note:** Some comment replies may appear under "outdated" sections because the code has been updated. All issues have been resolved.
